README.md
# App::Droplets

Create, destroy, inspect, and log on to your droplets from the command line

[![Build Status](https://travis-ci.org/sestaton/p5-app-droplets.svg?branch=master)](https://travis-ci.org/sestaton/p5-app-droplets)

**ABOUT**

This is a command-line tool for [DigitalOcean's](https://www.digitalocean.com/?refcode=c4cc062482a8) REST API called `droplets`. After setting up an API key on the [DigitalOcean](https://www.digitalocean.com/community/tutorials/how-to-use-the-digitalocean-api-v2) website, you can check on all your existing droplets, query the available droplet sizes and prices by region, create new droplets, or log on to droplets (either existing droplets or ones you just created) without leaving the command line. You can also destroy one or all your droplets in one command.

**INSTALLATION**

The following command will install the `droplets` application (note that this requires [git](http://git-scm.com/)):

    curl -sL cpanmin.us | perl - https://github.com/sestaton/digitalocean-cli.git

Alternatively, download the latest [release](https://github.com/sestaton/p5-app-droplets/releases) and run the following command in the top directory:

    perl Makefile.PL

If any Perl dependencies are listed after running this command, install them through the CPAN shell or any method you like. Then build and install the package.

    perl Makefile.PL
    make 
    make test
    make install

**USAGE**

The first time you run the tool you will be prompted for your API key (there is a link to a guide for setting up this token provided above in the **ABOUT** section of this document), so have that ready. Note that you also need to have your ssh keys set up with DigitalOcean. A configuration file (called ".droplets") will be stored in your home directory so you will not be prompted again. Once your key is stored, you are ready to go!

 Examples:

Let's see get some basic information.

    $ droplets --available sizes
    Name	Disk_space	Memory	CPUs	Cost_per_month	Cost_per_hour	Regions
    512mb	20	512	1	5	0.00744	nyc1,sgp1,ams1,sfo1,nyc2,lon1,nyc3,ams3,ams2,fra1
    1gb	30	1024	1	10	0.01488	nyc2,sgp1,ams1,sfo1,lon1,nyc3,ams3,nyc1,ams2,fra1
    2gb	40	2048	2	20	0.02976	nyc2,sfo1,ams1,sgp1,lon1,nyc3,ams3,nyc1,ams2,fra1
    4gb	60	4096	2	40	0.05952	nyc2,ams1,sgp1,lon1,nyc3,ams3,nyc1,ams2,sfo1,fra1
    8gb	80	8192	4	80	0.11905	nyc2,sgp1,ams1,nyc1,lon1,nyc3,ams3,ams2,sfo1,fra1
    16gb	160	16384	8	160	0.2381	sgp1,nyc1,nyc3,lon1,nyc2,ams1,ams3,ams2,sfo1,fra1
    32gb	320	32768	12	320	0.47619	nyc2,sgp1,nyc1,lon1,ams3,nyc3,ams2,sfo1,fra1
    48gb	480	49152	16	480	0.71429	sgp1,nyc1,lon1,nyc2,ams3,nyc3,ams2,sfo1,fra1
    64gb	640	65536	20	640	0.95238	sgp1,nyc1,nyc2,lon1,ams3,nyc3,ams2,sfo1,fra1

In the above output we can see the options available in different regions, along with the prices. Now, find out the available images..

    $ droplets --available images
    Distribution	ID	Name
    CoreOS	12789350	723.3.0 (beta)
    CentOS	6372321	5.10 x64
    CentOS	6372425	5.10 x32
    Debian	6372581	6.0 x64
    Debian	6372662	6.0 x32
    Fedora	9640922	21 x64
    FreeBSD	10144573	10.1
    Ubuntu	10321756	12.04.5 x64
    Ubuntu	10321777	12.04.5 x32
    Debian	10322059	7.0 x64
    Debian	10322378	7.0 x32
    CentOS	10322623	7 x64
    Fedora	12065782	22 x64
    Ubuntu	12658446	15.04 x64
    Ubuntu	12660649	15.04 x32
    Debian	12778278	8.1 x64
    Debian	12778337	8.1 x32
    CoreOS	13068283	723.3.0 (stable)
    Ubuntu	13089493	14.04 x64
    Ubuntu	13089823	14.04 x32

Now, let's create a droplet (in this case, Fedora 22)

    $ droplets --imageid 12065782 --create
    ---------------------------------------------------
    Created Droplet:         6588509
    Droplet name:            187ebaa6-42ac-11e5-bd97-de968ce03bd5
    Distribution:            Fedora 22 x64
    Region:                  San Francisco 1
    Disk:                    20
    Memory:                  512mb
    CPUs:                    1
    Server IP:               
    Creation time (seconds): 1
    ---------------------------------------------------

You can give the server a name, or just let a default be chosen, as above. If you want something you can remember, then specify a server name when you create the droplet. We should be able to see this droplet running if we check...

    $ droplets --running
    Running Droplets: 
    Droplet 187ebaa6-42ac-11e5-bd97-de968ce03bd5 has id 6588509, IP address 198.199.102.123, Gateway 198.199.102.1, and Netmask 255.255.255.0

You can log on to an existing server from the command line, or create one and log on. Here, we just use the defaults...

    $ droplets --create --login
    Droplet name:            908c31cf-42af-11e5-8d52-9bf6c9a52cfe
    Distribution:            CentOS 7 x64
    Region:                  San Francisco 1
    Disk:                    20
    Memory:                  512mb
    CPUs:                    1
    Server IP:               104.236.151.89
    Creation time (seconds): 1
    ---------------------------------------------------
    [root@908c31cf-42af-11e5-8d52-9bf6c9a52cfe ~]# ls
    [root@908c31cf-42af-11e5-8d52-9bf6c9a52cfe ~]# exit
    logout
    Connection to 104.236.151.89 closed.

You will notice the default is to log on as the root user. For images you plan to keep around, it is probably best to create regular user accounts using whatever methods apply for the Linux OS you created. Then, you can log on to the image using that user account instead of root:

    $ droplets --user evan --password somesecurepass --serverid 6588509 --logon
    [evan@908c31cf-42af-11e5-8d52-9bf6c9a52cfe ~]$

One final note about logging on to your droplets: you can use `--login`, `--logon` or `-l' to achieve the same purpose. I find this easier to work with than trying to remember the name of the command or having to look it up each time.

We can destroy a single droplet...

    $ droplets --destroy --serverid 6588509
    ====> Before destroy...
    Running Droplets:
    Droplet 187ebaa6-42ac-11e5-bd97-de968ce03bd5 has id 6588509, IP address 198.199.102.123, Gateway 198.199.102.1, and Netmask 255.255.255.0
    ====> After destroy...
    No running droplets.

Or, destroy all running droplets in one go. This convenient when you have a bunch of images that are no longer needed.

    $ droplets --destroyall
    ====> Before destroy...
    Running Droplets: 
    Droplet 908c31cf-42af-11e5-8d52-9bf6c9a52cfe has id 6588977, IP address 104.236.151.89, Gateway 104.236.128.1, and Netmask 255.255.192.0
    ====> After destroy...
    No running droplets.    


**SUPPORT AND DOCUMENTATION**

After installation, you can find documentation for App::Droplets with the `perldoc` command.

    perldoc droplets

The documentation can also be accessed by specifying the manual option with `droplets -m` or `droplets --man`. The `droplets` program will also print a diagnostic help message when executed with no arguments. **NB:** I mostly used long versions of options in the examples for clarity, but typing `droplets` will show the short versions of the options.

**ISSUES**

Report any issues or feature requests at the [issue tracker](https://github.com/sestaton/p5-app-droplets/issues).

**ATTRIBUTION**

This project uses the [Webservice::DigitalOcean](https://metacpan.org/pod/WebService::DigitalOcean) library to access the DigitalOcean APIv2.

**TODO**

- Add the option to change the defaults and store them in the configuration.

**LICENSE**

The MIT License should included with the project. If not, it can be found at: http://opensource.org/licenses/mit-license.php

Copyright (C) 2015-2016 S. Evan Staton

Changes
Revision history for App::Droplets

Version	Date	Location
0.07.4	12/14/2016	Vancouver, BC

- Bug fix when creating a droplet and logging on to it (option handling was
  modified in v0.07.3 requiring a server ID when logging on).

0.07.3	12/08/2016	Vancouver, BC

- Bug fix for logging on a regular user and not getting a prompt back.
- Add better option handling and error messages.

0.07.2	11/30/2016	Vancouver, BC

- Bug fix for logging on to droplet when creating an image.
- Bug fix for processing options in the correct order.
- Add alias for --login option (called --logon) to avoid having to retype commands due to
  this minor typo.
- Refactor code to log on to droplet by not having the same code repeated for separate options.

0.07.1	08/10/2016	Vancouver, BC

- Bug fix for when no arguments are given the program should generate a help message.
- Related to the above, handling multiple server IDs was not being evaluated correctly.

0.07.0	08/09/2016	Vancouver, BC

- Add option to take multiple server IDs with --destroy option. This was already in place
  if the IDs were joined by a comma, now this works with a list and is documented.
- Fix POD format with example output fields not being separated or readable. 

0.06 	03/11/2016	Vancouver, BC

- Add option for logging on to server as a specific user (other than root).

0.05	01/26/2016	Vancouver, BC

- Fix bug with wrong type specification for region and size parameters.
- Add method to validate parameters for creating a droplet.

0.04	01/23/2016	Vancouver, BC

- Fix bug with properly authenticating user token.

0.03	01/20/2016	Vancouver, BC

- Add test for authentication of API token prior to making
  any method calls. This is necessary because the object creation
  will succeed with an invalid token.
  
0.02	08/17/2015	Vancouver, BC

- Update build/install process with latest API version on CPAN,
  instead of building it from github.

0.01	08/14/2015	Vancouver, BC

- Initial release.	



Makefile.PL
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

if (eval ($ExtUtils::MakeMaker::VERSION) <= 6.55) {
    print "\n[ERROR]: ExtUtils::MakeMaker >= 6.5503 is required to build the package correctly.\n";
    print "The installed version is $ExtUtils::MakeMaker::VERSION.\n";
    print "Install the latest ExtUtils::MakeMaker and try again. Exiting.\n";
    exit(1);
}

WriteMakefile(
    NAME             => 'App::Droplets',
    AUTHOR           => q{S. Evan Staton <statonse@gmail.com>},
    VERSION_FROM     => 'bin/droplets',
    LICENSE          => 'MIT',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.010,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 6.5503,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
	'WebService::DigitalOcean' => 0.024,
	'Config::Tiny'             => 0,
	'UUID::Tiny'               => 0,
	'Type::Tiny'               => 0,
	'Expect'                   => 0,
    },
    INST_SCRIPT => 'blib/bin',
    EXE_FILES   => [ 'bin/droplets' ],
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    test  => { TESTS => 't/*.t' },
);



lib/App/Droplets.pm
package App::Droplets;

use 5.010;
use File::Spec;
use File::Basename;
use Expect;
use Config::Tiny;
use UUID::Tiny qw(:std);
use WebService::DigitalOcean;
#use Data::Dump::Color;

our $VERSION = '0.07.4';

#
# methods
#
sub new {
    my ($class, %args) = @_;

    my $self = bless {}, $class;
    $self->{opts} = $args{opts};

    return $self;
}

sub run {
    my $self = shift;
    my $opt  = $self->{opts};

    $opt = $self->configure($opt);
    my $token = $self->get_token($opt);
 
    my $do_obj = WebService::DigitalOcean->new({ token => $token });

    # make sure the token is valid before performing any method calls
    my $auth = $self->_confirm_authentication($do_obj);
    unless (defined $auth && $auth eq 'success') {
	say "\nERROR: Could not authenticate API token: $token\nCheck that the token is valid. Exiting.\n";
	exit(1);
    }
    
    if ($opt->{available}) {
	$self->report_available($do_obj, $opt->{available});
    }
    
    if ($opt->{running}) {
	$self->get_droplets($do_obj);
	exit(0);
    }

    my $server_addr = $self->create_droplet($do_obj, $config, $opt) if $opt->{create};

    if ($opt->{login} || $opt->{logon}) {
	my ($cmd, $serverid);
	if (@{$opt->{serverid}}) { 
	    if (@{$opt->{serverid}} == 1) {
		$serverid = shift @{$opt->{serverid}};
	    }
	    elsif (@{$opt->{serverid}} > 1) { 
		say "\nERROR: Only one server ID may be given with the '--login' option to avoid unexpected results. ".
		    "Check the '--serverid' arguments. Exiting.\n";
		exit(1);
	    }
	}
	
	my $user = $opt->{username} // 'root';
	my $ip = defined $serverid ? $self->get_address_for_droplet($do_obj, $serverid) : $server_addr;
	$cmd = sprintf "ssh $user@%s", $ip;
	my $ssh = Expect->new;
	$ssh->slave->clone_winsize_from(\*STDIN);

	if ($opt->{username}) {
	    $ssh->manual_stty(1);
	    $ssh->spawn($cmd);
	    #$ssh->slave->stty(qw(raw -echo)); # does not return prompt
	    $ssh->expect(10, 
			 [ 
			   'password: ',
			   sub { 
			       my $fh = shift;
			       say $fh $opt->{password};
			       exp_continue;
			   }
			   ],
		);
	    $ssh->interact();
	}
	else {
	    $ssh->spawn($cmd);
	    $ssh->interact();
	}
	$ssh->close();
    }

    if ($opt->{destroy}) {
	my @ids = @{$opt->{serverid}};
	$self->destroy_droplet($do_obj, \@ids);
    }

    if ($opt->{destroyall}) {
	$self->destroy_all($do_obj);
    }
}

sub configure {
    my $self = shift;
    my ($opt) = @_;

    my $conf_file = File::Spec->catfile($ENV{HOME}, '.droplets');
    if (-e $conf_file) {
	my $config = Config::Tiny->read( $conf_file, 'utf8' );
	return $opt if defined $config->{droplets}->{api_token};
    }
    elsif (defined $opt->{configfile}) {
	my $config = Config::Tiny->read( $opt->{configfile}, 'utf8' );
	return $opt if defined $config->{droplets}->{api_token};
    }
    else {
	print STDERR "\nApp::Droplets is not configured. Please provide your API token: ";
	my $token = <STDIN>;
	chomp $token;
	open my $out, '>', $conf_file;
	say $out "[droplets]";
	say $out "api_token = $token";
	close $out;

	$opt->{configfile} = $conf_file;
	return $opt;
    }
}

sub get_token {
    my $self = shift;
    my ($opt) = @_;

    $opt->{configfile} //= File::Spec->catfile($ENV{HOME}, '.droplets');
    my $config = Config::Tiny->read( $opt->{configfile}, 'utf8' );

    unless (defined $config->{droplets}->{api_token}) {
	say "\nERROR: The configuration is not formatted correctly. Exiting.\n";
	exit(1);
    }
    
    return $config->{droplets}->{api_token};
}

sub destroy_all {
    my $self = shift;
    my ($do_obj) = @_;

    my $droplets = $do_obj->droplet_list();

    my @ids;
    if ($droplets->{meta}{total} > 0) {
	for my $drop (@{$droplets->{content}}) {
	    if (defined $drop) {
		for my $ip (@{$drop->{networks}{v4}}) {
		    push @ids, $drop->{id};
		}
	    }
	}
    }

     $self->destroy_droplet($do_obj, \@ids);
}

sub destroy_droplet {
    my $self = shift;
    my ($do_obj, $ids) = @_;

    unless (@$ids) {
	say "\nERROR: Need at least 1 server ID to call destroy method. Exiting.\n";
	exit(1);
    }
    
    say "====> Before destroy...";
    $self->get_droplets($do_obj);
    for my $id (@$ids) {
	if (defined $id) {
	    $do_obj->droplet_delete($id);
	}
    }
    sleep 10;
    say "====> After destroy...";
    $self->get_droplets($do_obj);
}

sub create_droplet {
    my $self = shift;
    my ($do_obj, $config, $opt) = @_;

    my $ssh_id = $self->get_keys($config);
    $opt->{region}  //= 'sfo1';        # San Francisco
    $opt->{size}    //= '512mb';       # 512MB
    $opt->{name}    //= create_uuid_as_string(UUID_V1);
    $opt->{imageid} //= 17384153;      # CentOS 7.2 x64
    
    my $t0 = time;
    my $droplet = $do_obj->droplet_create({
	name     => $opt->{name},
	size     => $opt->{size},
	image    => $opt->{imageid},
	region   => $opt->{region},
	ssh_keys => [ $ssh_id ],
    });
    my $t1 = time;

    sleep 5;
    my $success = $self->_check_droplet_params($droplet);
    if ($success) {
	my $server = $do_obj->droplet_get($droplet->{content}{id});
	my $ip = $self->get_address_for_droplet($do_obj, $server->{content}{id});
        
	say "---------------------------------------------------";
	say "Created Droplet:         ", $server->{content}{id}; 
	say "Droplet name:            ", $server->{content}{name}; 
	say "Distribution:            ", $server->{content}{image}{distribution}.q{ }.$server->{content}{image}{name};
	say "Region:                  ", $server->{content}{region}{name};
	say "Disk:                    ", $server->{content}{size}{disk};
	say "Memory:                  ", $server->{content}{size_slug};
	say "CPUs:                    ", $server->{content}{vcpus};
	say "Server IP:               ", $ip;
	say "Creation time (seconds): ", $t1-$t0;
	say "---------------------------------------------------";

	return $server->{content}{networks}{v4}[0]{ip_address};
    }
}

sub report_available {
    my $self = shift;
    my ($obj, $available) = @_;

    my %disp = (
	'sizes'   => sub { $self->get_sizes($obj)   },
	'images'  => sub { $self->get_images($obj)  },
	'regions' => sub { $self->get_regions($obj) },
    );

    if (defined $available && exists $disp{$available}) {
	$disp{$available}->();
    }
    else {
	say STDERR "\nERROR: '$available' is not recognized. ".
	    "Argument must be one of: sizes, images, or regions. Exiting.\n";
	exit(1);
    }
    
    exit(0);
}
	
	
sub get_sizes {
    my $self = shift;
    my ($do_obj) = @_;
    
    my $sizes = $do_obj->size_list();
    
    say join "\t", "Name", "Disk_space", "Memory", "CPUs", "Cost_per_month", "Cost_per_hour", "Regions";
    for my $s (@{$sizes->{content}}) {
	say join "\t", $s->{slug}, $s->{disk}, $s->{memory}, $s->{vcpus}, 
	    $s->{price_monthly}, $s->{price_hourly}, join ',', @{$s->{regions}};
    }

    exit(0);
}

sub get_regions {
    my $self = shift;
    my ($do_obj) = @_;

    my $region_id;
    my $regions = $do_obj->region_list();
    
    say join "\t", "Name", "Slug", "Sizes";
    for my $r (@{$regions->{content}}) {
	say join "\t", $r->{name}, $r->{slug}, join ',', @{$r->{sizes}};
    }

    exit(0);
}

sub get_images {
    my $self = shift;
    my ($do_obj) = @_;
    
    my $image_id;
    my $images = $do_obj->image_list();

    say join "\t", "Distribution", "ID", "Name";
    for my $img (@{$images->{content}}) {
	say join "\t", $img->{distribution}, $img->{id}, $img->{name};
    }

    exit(0);
}

sub get_keys {
    my $self = shift;
    my ($config) = @_;

    ## this creates a new key fingerprint
    my $key = File::Spec->catfile($ENV{HOME}, '.ssh', 'id_rsa.pub');
    my $cmd = "ssh-keygen -lf $key";
    my $out = qx($cmd);
    chomp $out;
    my ($bits, $fingerprint, $identity) = split /\s+/, $out;

    return $fingerprint;
}

sub get_address_for_droplet {
    my $self = shift;
    my ($do_obj, $id) = @_;

    my $sid = ref($id) ? shift(@$id) : $id;

    my $server = $do_obj->droplet_get($sid);
    my $ip = $server->{content}{networks}{v4}[0]{ip_address};
    
    return $ip;
}

sub get_droplets {
    my $self = shift;
    my ($do_obj) = @_;

    my $droplets = $do_obj->droplet_list();

    my @ips;
    if ($droplets->{meta}{total} > 0) {
	say "Running Droplets: ";
	for my $drop (@{$droplets->{content}}) {
	    if (defined $drop) {
		for my $ip (@{$drop->{networks}{v4}}) {
		    printf "Droplet %s has id %s, IP address %s, Gateway %s, and Netmask %s\n", 
	            $drop->{name}, $drop->{id}, $ip->{ip_address}, $ip->{gateway}, $ip->{netmask};
		}
	    }
	}
    }
    else {
	say "No running droplets.";
    }
}

sub _confirm_authentication {
    my $self = shift;
    my ($do_obj) = @_;

    my $content = $do_obj->region_list->{content};

    if (ref($content) eq 'HASH') {
	if (defined $content->{id} && $content->{id} eq 'unauthorized') { # && 
	    #defined $content->{message} && $content->{message} eq 'Unable to authenticate you.') {
	    return undef;
	}
    }
    else {
	return 'success';
    }
}

sub _check_droplet_params {
    my $self = shift;
    my ($do_obj) = @_;

    if ($do_obj->{content}{id} eq 'unprocessable_entity') {
	say STDERR "\nERROR: There appears to be an error with the input parameters.\n";
	say STDERR "Here is the message: ",$do_obj->{content}{message};
	exit(1);
    }
    else {
	return 1;
    }
}

1;

__END__

=head1 NAME 
                                                                       
App::Droplets - Create, destroy, inspect, and log on to your droplets from the command line

=head1 SYNOPSIS    

  ## get the available image sizes
  droplets --available sizes

  # get the available distributions
  droplets --available images
    Distribution  ID          Name
    CoreOS        12789350    723.3.0 (beta)
    CentOS        6372321     5.10 x64
    CentOS        6372425     5.10 x32
    Debian        6372581     6.0 x64
    Debian        6372662     6.0 x32
    Fedora        9640922     21 x64
    FreeBSD       10144573    10.1
    Ubuntu        10321756    12.04.5 x64
    Ubuntu        10321777    12.04.5 x32
    Debian        10322059    7.0 x64
    Debian        10322378    7.0 x32
    CentOS        10322623    7 x64
    Fedora        12065782    22 x64
    Ubuntu        12658446    15.04 x64
    Ubuntu        12660649    15.04 x32
    Debian        12778278    8.1 x64
    Debian        12778337    8.1 x32
    CoreOS        13068283    723.3.0 (stable)
    Ubuntu        13089493    14.04 x64
    Ubuntu        13089823    14.04 x32

  ## Create a droplet (in this case, Fedora 22)
  droplets --imageid 12065782 --create

  ## Check all running droplets
  droplets --running

  ## destroy one specific droplet
  droplets --destroy --serverid 6588509

  ## destroy numerous droplets in one go                                                                                                droplets --destroy --serverid 6588509 6588510 6588511

  ## Create a droplet, with defaults, and log on to it
  droplets --create --login

  ## Destroy all running droplets
  droplets --destroyall

=head1 DESCRIPTION
     
 This is a command-line tool for DigitalOcean's REST API called droplets. After setting up an API key on the DigitalOcean website, you can check on all your existing droplets, query the available droplet sizes and prices by region, create new droplets, or log on to droplets (either existing droplets or ones you just created) without leaving the command line. 
=head1 LICENSE
 
The MIT License should included with the project. If not, it can be found at: http://opensource.org/licenses/mit-license.php

Copyright (C) 2015-2016 S. Evan Staton

=cut
